compiler/rustc_resolve/src/build_reduced_graph.rs

```diff
@@ -93,59 +93,68 @@ impl<'a> Resolver<'a> {
-        // Walk up the tree until we hit a module or the crate root.
-        while parent_id.index != CRATE_DEF_INDEX
-            && self.cstore().def_kind(parent_id) != DefKind::Mod
-        {
-            let parent_def_key = self.cstore().def_key(parent_id);
-            parent_id.index = parent_def_key.parent.expect("failed to get parent for module");
+    /// stopping at the first encountered module.
+    /// Parent block modules for arbitrary def-ids are not recorded for the local crate,
+    /// and are not preserved in metadata for foreign crates, so block modules are never returned
+    /// by this function, which may or may not be correct depending on the use case.
+    fn get_nearest_non_block_module(&mut self, mut def_id: DefId) -> Module<'a> {
+        loop {
+            match self.get_module(def_id) {
+                Some(module) => return module,
+                None => {
+                    def_id.index =
+                        self.def_key(def_id).parent.expect("non-root `DefId` without parent")
+                }
+            }
         }
-        self.get_module(parent_id)
     }
 
-    pub fn get_module(&mut self, def_id: DefId) -> Module<'a> {
-        // Cache module resolution
-        if let Some(module) = self.module_map.get(&def_id) {
-            return *module;
+    pub fn expect_module(&mut self, def_id: DefId) -> Module<'a> {
+        self.get_module(def_id).expect("argument `DefId` is not a module")
+    }
+
+    /// If `def_id` refers to a module (in resolver's sense, i.e. a module item, crate root, enum,
+    /// or trait), then this function returns that module's resolver representation, otherwise it
+    /// returns `None`.
+    /// FIXME: `Module`s for local enums and traits are not currently found.
+    crate fn get_module(&mut self, def_id: DefId) -> Option<Module<'a>> {
+        if let module @ Some(..) = self.module_map.get(&def_id) {
+            return module.copied();
         }
 
-        assert!(!def_id.is_local());
-        let (name, parent) = if def_id.index == CRATE_DEF_INDEX {
-            // This is the crate root
-            (self.cstore().crate_name(def_id.krate), None)
-        } else {
-            let def_key = self.cstore().def_key(def_id);
-            let name = def_key
-                .disambiguated_data
-                .data
-                .get_opt_name()
-                .expect("given a DefId that wasn't a module");
-
-            let parent = Some(self.nearest_parent_mod(def_id));
-            (name, parent)
-        };
+        if !def_id.is_local() {
+            let def_kind = self.cstore().def_kind(def_id);
+            match def_kind {
+                DefKind::Mod | DefKind::Enum | DefKind::Trait => {
+                    let def_key = self.cstore().def_key(def_id);
+                    let parent = def_key.parent.map(|index| {
+                        // Block modules are simply not kept in metadata, so we have to assume that
```

Comment:
Not keeping them shouldn't affect stable rust because nothing from these blocks is supposed to be reachable from other crates. It would be good if we didn't have to keep them.

In theory, if a macro can be both `#[macro_export]` and use def-site hygiene, then it will be able to use items from its otherwise unreachable unnamed def-site block. However, this case is unstable and we can prohibit it if necessary.

Not keeping this data can also make rustdoc link resolution less precise in corner cases.
If a public impl resides in a block, then links in its doc comments should ideally be resolved from inside that block, but it will be performed from its nearest normal module instead.

(I need to check what else `ModuleData::parent` is used for.)

Reply:
>I need to check what else `ModuleData::parent` is used for.

Yep, it's either checks like `parent.is_some()` where block modules don't matter, or some kind of resolution in that module's scope where the parent for foreign modules only matters in case of macros 2.0 with def-site hygiene.

Reply:
I've reworded the comments to be a bit more reassuring.